plugins/usbdmx/ShowJockey.cpp

```diff
@@ -0,0 +1,393 @@
+  int endpoint;
+  int max_packet_size;
+  if (!LocateInterface(m_adaptor,
+                       m_usb_device,
+                       &interface_number,
+                       &endpoint,
+                       &max_packet_size)) {
+    return false;
+  }
+
+  bool ok = m_adaptor->OpenDeviceAndClaimInterface(m_usb_device,
+                                                   interface_number,
+                                                   &usb_handle);
+  if (!ok) {
+    return false;
+  }
+
+  std::auto_ptr<ShowJockeyThreadedSender> sender(
+      new ShowJockeyThreadedSender(m_adaptor,
+                                   m_usb_device,
+                                   usb_handle,
+                                   max_packet_size,
+                                   endpoint));
+  if (!sender->Start()) {
+    return false;
+  }
+
+  m_sender.reset(sender.release());
+  return true;
+}
+
+bool SynchronousShowJockey::SendDMX(const DmxBuffer &buffer) {
+  return m_sender.get() ? m_sender->SendDMX(buffer) : false;
+}
+
+// ShowJockeyAsyncUsbSender
+// -----------------------------------------------------------------------------
+class ShowJockeyAsyncUsbSender : public AsyncUsbSender {
+ public:
+  ShowJockeyAsyncUsbSender(LibUsbAdaptor *adaptor,
+                           libusb_device *usb_device,
+                           int endpoint,
+                           int max_packet_size_out,
+                           libusb_device_handle *handle)
+                           : AsyncUsbSender(adaptor, usb_device),
+                             m_endpoint(endpoint),
+                             m_max_packet_size_out(max_packet_size_out) {
+    m_usb_handle = handle;
+    m_tx_frame = NULL;
+  }
+
+  ~ShowJockeyAsyncUsbSender() {
+    CancelTransfer();
+    delete[] m_tx_frame;
+  }
+
+  libusb_device_handle* SetupHandle() {
+    return m_usb_handle;
+  }
+
+  bool PerformTransfer(const DmxBuffer &buffer) {
+    uint16_t nb_sequence = DMX_MAX_SLOT_NUMBER / (m_max_packet_size_out - 2);
+    int final_size = DMX_MAX_SLOT_NUMBER + 2 * nb_sequence;
+    if (m_tx_frame == NULL) {
+      m_tx_frame = new uint8_t[final_size]();
+    }
+
+    uint8_t *current_buffer = new uint8_t[DMX_MAX_SLOT_NUMBER]();
+    unsigned int dmx_max_slot_number = DMX_MAX_SLOT_NUMBER;
+    buffer.Get(current_buffer, &dmx_max_slot_number);
+    uint8_t *p_current_buffer = current_buffer;
+    uint8_t *p_final_buffer = m_tx_frame;
+    int to_write_size = m_max_packet_size_out - 2;
+    uint16_t written_size = 0;
+    for (int i = 0; i <= nb_sequence; ++i) {
+      memcpy(p_final_buffer, &written_size, 2);
+      p_final_buffer += 2;
+      int need_to_write_size = DMX_MAX_SLOT_NUMBER - written_size;
+      int will_write_size;
+      if (to_write_size < need_to_write_size) {
+        will_write_size = to_write_size;
+      } else {
+        will_write_size = need_to_write_size;
+      }
+
+      memcpy(p_final_buffer, p_current_buffer, will_write_size);
```

Comment:
Can you not use http://docs.openlighting.org/ola/doc/latest/classola_1_1_dmx_buffer.html#ad78488db3fec3fd5e47803d77ef22e5e ?